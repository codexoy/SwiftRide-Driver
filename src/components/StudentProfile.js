import React, { Fragment } from "react"
import { View, StyleSheet, Text, Image, TouchableOpacity } from "react-native"
import { strings } from "../i18n"
import UserWithPic from "./UserWithPic"
import ShadowRect from "./ShadowRect"
import { MAIN_PADDING } from "../consts"
import { Icon } from "react-native-elements"
import TopicsList from "./TopicsList"
import StudentPayments from "./StudentPayments"
import { getPayments } from "../actions/lessons"
import StudentNextLessonView from "./StudentNextLessonView"

export default class StudentProfile extends React.Component {
	constructor(props) {
		super(props)
		let student = this.props.user
		if (this.props.navigation.getParam("student")) {
			// if we are in other student's profile
			student = this.props.navigation.getParam("student")
		}
		let isTeacher = false
		if (this.props.user.hasOwnProperty("teacher_id")) {
			isTeacher = true
		}
		this.state = {
			student,
			allTopics: [],
			combinedTopics: [],
			payments: [],
			nextLesson: "",
			isTeacher
		}

		this._getTopics()
		if (isTeacher) {
			this._getNextLesson()
			this._getPayments()
		}
	}

	_getNextLesson = async () => {
		const now = new Date().toISOString()
		const resp = await this.props.fetchService.fetch(
			"/lessons/?limit=1&is_approved=true&date=ge:" +
				now +
				"&student_id=" +
				this.state.student.student_id,
			{ method: "GET" }
		)
		if (resp.json["data"].length == 0) return
		this.setState({
			nextLesson: resp.json["data"][0]
		})
	}

	_getPayments = async () => {
		const payments = await getPayments(
			this.props.fetchService,
			false,
			`student_id=${this.state.student.student_id}`
		)
		this.setState({
			payments: payments.payments
		})
	}
	_getTopics = async () => {
		const resp = await this.props.fetchService.fetch(
			`/student/${this.state.student.student_id}/topics`,
			{ method: "GET" }
		)
		const combinedTopics = [].concat.apply(
			[],
			Object.entries(resp.json["data"])
				.map(x => {
					if (x[0] != "new") return x[1]
				})
				.filter(x => x != undefined)
		)
		this.setState({
			allTopics: resp.json["data"],
			combinedTopics: combinedTopics
		})
	}
	render() {
		const { student } = this.state

		let teacherView
		if (this.state.isTeacher) {
			// teacher is logged in, show next lesson and payments
			teacherView = (
				<Fragment>
					<ShadowRect style={styles.rect}>
						<Text style={styles.rectTitle} testID="schedule">
							{strings("teacher.home.next_lesson")}
						</Text>
						<StudentNextLessonView
							testID="lessonRowTouchable"
							lesson={this.state.nextLesson}
						/>
					</ShadowRect>
					<ShadowRect style={styles.rect}>
						<Text style={styles.rectTitle} testID="schedule">
							{strings("teacher.students.balance")}
						</Text>
						<StudentPayments
							sum={this.state.student.balance}
							payments={this.state.payments.slice(0, 2)}
						/>
					</ShadowRect>
				</Fragment>
			)
		}
		return (
			<View style={styles.container}>
				<View style={styles.header}>
					<UserWithPic
						name={student.user.name}
						extra={
							<Text>
								{student.new_lesson_number}{" "}
								{strings("student_profile.lessons")}
							</Text>
						}
						width={54}
						height={54}
					/>
					<View style={styles.badges}>
						<Image
							style={styles.badge}
							source={{
								uri:
									"https://images.unsplash.com/photo-1535643302794-19c3804b874b?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2134&q=80"
							}}
						/>
					</View>
				</View>
				{teacherView}
				<ShadowRect style={styles.rect}>
					<View style={{ flexDirection: "row" }}>
						<Text testID="monthlyAmount" style={styles.rectTitle}>
							{strings("teacher.new_lesson.topics", {
								topics: this.state.allTopics
							})}
						</Text>

						<View
							style={{
								flex: 1,
								marginLeft: "auto",
								alignItems: "flex-end"
							}}
						>
							<TouchableOpacity
								onPress={() =>
									this.props.navigation.navigate("Topics", {
										topics: this.state.allTopics
									})
								}
							>
								<Icon
									name="arrow-back"
									type="material"
									size={20}
								/>
							</TouchableOpacity>
						</View>
					</View>
					<TopicsList topics={this.state.combinedTopics} />
				</ShadowRect>
			</View>
		)
	}
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		marginLeft: MAIN_PADDING,
		marginRight: MAIN_PADDING,
		marginTop: 20
	},
	header: {
		maxHeight: 60,
		flexDirection: "row",
		alignItems: "center"
	},
	rect: {
		marginTop: 20
	},
	rectTitle: {
		fontWeight: "bold",
		color: "rgb(121, 121, 121)"
	},
	badge: {
		height: 24,
		width: 24,
		borderRadius: 12,
		marginLeft: 12,
		marginTop: -2
	},
	badges: {
		marginLeft: "auto",
		marginTop: -6
	}
})
