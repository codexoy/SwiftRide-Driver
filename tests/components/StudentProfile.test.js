import React from "react"

import renderer from "react-test-renderer"
import FetchService from "../../src/services/Fetch"
import StudentProfile from "../../src/components/StudentProfile"

const fetchService = new FetchService()
const student = {
	student_id: 1,
	balance: 900,
	user: { name: "test" },
	my_teacher: { user: {} }
}
const navigation = {
	navigate: jest.fn(),
	getParam: jest.fn(),
	addListener: jest.fn()
}
const topics = {
	new: [],
	in_progress: [{ id: 1, title: "test1" }],
	finished: [{ id: 2, title: "test2" }, { id: 3, title: "test3" }]
}
const payments = [
	{ id: 1, amount: 100, details: "test", pdf_link: "http://test.com" },
	{ id: 1, amount: 100, details: "test" }
]
const nextLesson = [{ id: 1 }]
fetch.mockResponseSuccess(
	JSON.stringify({
		data: topics
	})
)
jest.useFakeTimers()
describe("StudentProfile", () => {
	test("view renders correctly -> teacher's view", () => {
		fetch.mockResponseSuccess(
			JSON.stringify({
				data: nextLesson
			})
		)
		fetch.mockResponseSuccess(
			JSON.stringify({
				data: payments
			})
		)
		const tree = renderer
			.create(
				<StudentProfile
					user={{
						...student,
						teacher_id: 1
					}}
					navigation={navigation}
					fetchService={fetchService}
				/>
			)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})

	test("view renders correctly -> student's view", () => {
		const tree = renderer
			.create(
				<StudentProfile
					user={student}
					navigation={navigation}
					fetchService={fetchService}
				/>
			)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})
})
