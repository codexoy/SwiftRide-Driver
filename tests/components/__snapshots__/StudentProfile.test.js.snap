// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentProfile view renders correctly -> student's view 1`] = `
<ScrollViewMock
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "marginTop": 20,
        "paddingLeft": 26,
        "paddingRight": 26,
      }
    }
  >
    <ContactPopup
      onPress={[Function]}
      visible={false}
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "maxHeight": 60,
        }
      }
    >
      <UserWithPic
        extra={
          <View
            style={
              Object {
                "alignItems": "flex-start",
              }
            }
          >
            <Text>
               
              שיעורים
            </Text>
            <TouchableOpacity
              activeOpacity={0.2}
              onPress={[Function]}
            >
              <Text
                style={
                  Object {
                    "color": "rgb(12,116,244)",
                  }
                }
              >
                ערוך פרטים
              </Text>
            </TouchableOpacity>
          </View>
        }
        height={54}
        user={
          Object {
            "balance": 900,
            "my_teacher": Object {
              "user": Object {
                "id": 2,
                "name": "teacher",
              },
            },
            "name": "test",
            "student_id": 1,
          }
        }
        width={54}
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginLeft": "auto",
            "marginTop": -6,
          }
        }
      >
        <Text>
          המורה שלי:
        </Text>
        <FastImage
          resizeMode="cover"
          source={
            Object {
              "uri": "https://res.cloudinary.com/hsmkon3br/image/upload/c_thumb,g_face,h_80,w_80/v1554188012/678099-profile-filled-256.png",
            }
          }
          style={
            Object {
              "borderRadius": 12,
              "height": 24,
              "marginLeft": 12,
              "marginTop": -2,
              "width": 24,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginTop": 24,
        }
      }
    />
    <View
      style={
        Object {
          "marginTop": 20,
        }
      }
    />
    <ShadowRect
      style={Object {}}
    >
      <SimpleLoader />
    </ShadowRect>
  </View>
</ScrollViewMock>
`;

exports[`StudentProfile view renders correctly -> teacher's view 1`] = `
<ScrollViewMock
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "marginTop": 20,
        "paddingLeft": 26,
        "paddingRight": 26,
      }
    }
  >
    <ContactPopup
      onPress={[Function]}
      visible={false}
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "maxHeight": 60,
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "marginRight": 8,
          }
        }
      >
        <Themed.Icon
          name="arrow-forward"
          type="material"
        />
      </TouchableOpacity>
      <UserWithPic
        extra={
          <View
            style={
              Object {
                "alignItems": "flex-start",
              }
            }
          >
            <Text>
               
              שיעורים
            </Text>
            <TouchableOpacity
              activeOpacity={0.2}
              onPress={[Function]}
            >
              <Text
                style={
                  Object {
                    "color": "rgb(12,116,244)",
                  }
                }
              >
                ערוך פרטים
              </Text>
            </TouchableOpacity>
          </View>
        }
        height={54}
        user={
          Object {
            "balance": 900,
            "my_teacher": Object {
              "user": Object {
                "id": 2,
                "name": "teacher",
              },
            },
            "name": "test",
            "student_id": 1,
          }
        }
        width={54}
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginLeft": "auto",
            "marginTop": -6,
          }
        }
      >
        <Text>
          המורה שלי:
        </Text>
        <FastImage
          resizeMode="cover"
          source={
            Object {
              "uri": "https://res.cloudinary.com/hsmkon3br/image/upload/c_thumb,g_face,h_80,w_80/v1554188012/678099-profile-filled-256.png",
            }
          }
          style={
            Object {
              "borderRadius": 12,
              "height": 24,
              "marginLeft": 12,
              "marginTop": -2,
              "width": 24,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginTop": 24,
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "padding": 12,
          }
        }
      >
        <Themed.Icon
          name="phone"
          size={24}
          type="feather"
        />
        <Text
          style={
            Object {
              "marginTop": 8,
            }
          }
        >
          צור קשר
        </Text>
      </TouchableOpacity>
    </View>
    <ShadowRect
      style={
        Object {
          "marginTop": 20,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(121, 121, 121)",
            "fontWeight": "bold",
          }
        }
        testID="schedule"
      >
        שיעור הבא
      </Text>
      <StudentNextLessonView
        lesson={
          Object {
            "duration": 40,
            "id": 1,
          }
        }
        lessonPress={[Function]}
        loading={false}
        testID="lessonRowTouchable"
      />
      <LessonPopup
        item={
          Object {
            "duration": 40,
            "id": 1,
          }
        }
        navigation={
          Object {
            "addListener": [Function],
            "getParam": [Function],
            "navigate": [MockFunction],
          }
        }
        onPress={[Function]}
        testID="lessonPopup"
        visible={false}
      />
    </ShadowRect>
    <ShadowRect
      style={Object {}}
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(121, 121, 121)",
              "fontWeight": "bold",
            }
          }
        >
          מאזן כספי
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flex": 1,
              "marginRight": "auto",
            }
          }
        >
          <Themed.Icon
            name="arrow-back"
            size={20}
            type="material"
          />
        </View>
      </TouchableOpacity>
      <StudentPayment
        dispatch={
          [MockFunction] {
            "calls": Array [
              Array [
                [Function],
              ],
              Array [
                [Function],
              ],
              Array [
                [Function],
              ],
              Array [
                [Function],
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": null,
              },
              Object {
                "type": "return",
                "value": null,
              },
              Object {
                "type": "return",
                "value": null,
              },
              Object {
                "type": "return",
                "value": null,
              },
            ],
          }
        }
        loading={false}
        payments={
          Array [
            Object {
              "amount": 100,
              "details": "test",
              "id": 1,
              "pdf_link": "http://test.com",
            },
            Object {
              "amount": 100,
              "details": "test",
              "id": 1,
            },
          ]
        }
        sum={900}
        user={
          Object {
            "name": "teacher",
            "teacher_id": 1,
          }
        }
      />
    </ShadowRect>
    <ShadowRect
      style={Object {}}
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(121, 121, 121)",
              "fontWeight": "bold",
            }
          }
          testID="monthlyAmount"
        >
          נושאים
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flex": 1,
              "marginLeft": "auto",
            }
          }
        >
          <Themed.Icon
            name="arrow-back"
            size={20}
            type="material"
          />
        </View>
      </TouchableOpacity>
      <TopicsList
        topics={
          Array [
            Object {
              "id": 1,
              "title": "test1",
            },
            Object {
              "id": 2,
              "title": "test2",
            },
            Object {
              "id": 3,
              "title": "test3",
            },
          ]
        }
      />
    </ShadowRect>
  </View>
</ScrollViewMock>
`;
